core/lib/zksync_core/src/commitment_generator/mod.rs

```diff
@@ -0,0 +1,266 @@
+    commitment::{AuxCommitments, CommitmentCommonInput, CommitmentInput, L1BatchCommitment},
+    writes::{InitialStorageWrite, RepeatedStorageWrite, StateDiffRecord},
+    L1BatchNumber, ProtocolVersionId, StorageKey, H256,
+};
+use zksync_utils::h256_to_u256;
+
+mod metrics;
+
+const SLEEP_INTERVAL: Duration = Duration::from_millis(100);
+
+#[derive(Debug)]
+pub struct CommitmentGenerator {
+    connection_pool: ConnectionPool,
+    stop_receiver: watch::Receiver<bool>,
+}
+
+impl CommitmentGenerator {
+    pub fn new(connection_pool: ConnectionPool, stop_receiver: watch::Receiver<bool>) -> Self {
+        Self {
+            connection_pool,
+            stop_receiver,
+        }
+    }
+
+    async fn calculate_aux_commitments(
+        &self,
+        l1_batch_number: L1BatchNumber,
+        protocol_version: ProtocolVersionId,
+    ) -> anyhow::Result<AuxCommitments> {
+        let mut connection = self
+            .connection_pool
+            .access_storage_tagged("commitment_generator")
+            .await?;
+        let events_queue = connection
+            .blocks_dal()
+            .get_events_queue(l1_batch_number)
+            .await?
+            .context("Events queue is required for post-boojum batch")?;
+        let initial_bootloader_contents = connection
+            .blocks_dal()
+            .get_initial_bootloader_heap(l1_batch_number)
+            .await?
+            .context("Bootloader initial heap is missing")?;
+        drop(connection);
+
+        let events_commitment_task: JoinHandle<anyhow::Result<H256>> =
+            tokio::task::spawn_blocking(move || {
+                let latency = METRICS.events_queue_commitment_latency.start();
+                let events_queue_commitment =
+                    events_queue_commitment(&events_queue, protocol_version)
+                        .context("Events queue commitment is required for post-boojum batch")?;
+                latency.observe();
+
+                Ok(events_queue_commitment)
+            });
+
+        let bootloader_memory_commitment_task: JoinHandle<anyhow::Result<H256>> =
+            tokio::task::spawn_blocking(move || {
+                let latency = METRICS.bootloader_content_commitment_latency.start();
+                let bootloader_initial_content_commitment = bootloader_initial_content_commitment(
+                    &initial_bootloader_contents,
+                    protocol_version,
+                )
+                .context("Bootloader content commitment is required for post-boojum batch")?;
+                latency.observe();
+
+                Ok(bootloader_initial_content_commitment)
+            });
+
+        let events_queue_commitment = events_commitment_task
+            .await
+            .context("`events_commitment_task` failed")??;
+        let bootloader_initial_content_commitment = bootloader_memory_commitment_task
+            .await
+            .context("`bootloader_memory_commitment_task` failed")??;
+
```

Comment:
Does it make sense to spawn them in parallel?

Reply:
Yes, events commitment take around 500ms on mainnet at the moment and bootloader memory around 2s, so we 500ms by running in parallel